.github/workflows/_build-and-test-locally.yml

```diff
@@ -334,24 +344,127 @@ jobs:
         with:
           submodules: true
 
+      # Needed for detecting new and updated tests
+      - uses: actions/checkout@v4
+        if: github.event.pull_request.head.sha && inputs.test-selection == '' && matrix.pg_version == 'v17'
+        with:
+          submodules: false
+          clean: false
+          fetch-depth: 1000
+
+      - name: Cache poetry deps
+        if: github.event.pull_request.head.sha && inputs.test-selection == '' && matrix.pg_version == 'v17'
+        uses: actions/cache@v4
+        with:
+            path: ~/.cache/pypoetry/virtualenvs
+            key: v2-${{ runner.os }}-${{ runner.arch }}-python-deps-bookworm-${{ hashFiles('poetry.lock') }}
+
+      - name: Install Python deps
+        if: github.event.pull_request.head.sha && inputs.test-selection == '' && matrix.pg_version == 'v17'
+        shell: bash -euxo pipefail {0}
+        run: ./scripts/pysync
+
+      - name: Detect new and updated tests
+        id: select_tests_to_test
+        if: github.event.pull_request.head.sha && inputs.test-selection == '' && matrix.pg_version == 'v17'
+        env:
+          COMMIT_SHA: ${{ github.event.pull_request.head.sha || github.sha }}
+          BASE_SHA: ${{ github.event.pull_request.base.sha || github.sha }}
+        run: |
+          MAX_TESTS=3
+          CURRENT="${COMMIT_SHA}"
+          BASELINE="$(git merge-base $BASE_SHA $CURRENT)"
+          if [ -z $BASELINE ]; then
+            echo "Baseline commit for PR is not found, detection skipped."
+            exit 0
+          fi
+          true > /tmp/neon-pytest-new-tests.txt
+          true > /tmp/neon-pytest-updated-tests.txt
+          for t in $(git diff --name-only $BASELINE..$CURRENT test_runner/regress/); do
+            [ -f "$t" ] || continue
+            echo "Test file modified: $t"
+
+            # Get and compare two lists of items collected by pytest to detect new tests in the PR
+            { git show $BASELINE:"$t" || true; } > test_runner/regress/__temp__.py
+            { ./scripts/pytest --collect-only -q test_runner/regress/__temp__.py | sed '/^$/q; s/.*:://; s/\[.*//' | sort -u || true; } > /tmp/neon-pytest-tests-0.txt
+            cat /tmp/neon-pytest-tests-0.txt
+            { ./scripts/pytest --collect-only -q "$t" | sed '/^$/q; s/.*:://; s/\[.*//' | sort -u || true; } > /tmp/neon-pytest-tests-1.txt
+            cat /tmp/neon-pytest-tests-1.txt
+            comm -13 /tmp/neon-pytest-tests-0.txt /tmp/neon-pytest-tests-1.txt | sed "s|^|$t::|" >> /tmp/neon-pytest-new-tests.txt
+
+            # Detect pre-existing test functions updated in the PR
```

Comment:
I'd postpone adding these changes. Let's start with a smaller subset of features, i.e.
Retry all new tests N times.
And once we get used to that we can add the detection of changes in test blocks itself